File: 0664-strange-printer/0664-strange-printer.cpp
class Solution {
public:
    int dp[100][100];
    int solve(string s, int i, int j) {
        if(i > j) return 0;
        if(dp[i][j] != -1) return dp[i][j];
        int ans = solve(s, i, j - 1) + 1;
        for(int k = i; k<j; k++) {
            if(s[k] == s[j]) {
                ans = min(ans, solve(s, i, k-1) + solve(s, k, j-1));
            }
        }
        return dp[i][j] = ans;
    }

    int strangePrinter(string s) {
        int n = s.size();
        if(n == 0 || n == 1) return n;
        string str = "";
        str += s[0];
        for(int i = 1; i<n; i++) {
            if(s[i] == s[i-1]) continue;
            else str += s[i];
        }
        cout<<str<<endl;
        memset(dp, -1, sizeof(dp));
        return solve(str, 0, str.size() - 1);
    }
};

File: 0664-strange-printer/0664-strange-printer.py
class Solution(object):
    def solve(self, s, i, j, dp):
        if(i, j) not in dp:
            if i> j: return 0

            ans = self.solve(s, i, j - 1, dp) + 1
            for k in range(i, j):
                if s[k] == s[j]:
                    ans = min(ans, self.solve(s, i, k - 1, dp) + self.solve(s, k , j - 1, dp))
            dp[(i, j)] = ans;
        return dp[(i, j)]

    def strangePrinter(self, s):
        s = re.sub(r'(.)\1*', r'\1', s)
        dp = {}
        return self.solve(s, 0, len(s) - 1, dp)